azurelinuxagent/ga/agent_update_handler.py

```diff
@@ -182,12 +182,26 @@ def run(self, goal_state, ext_gs_updated):
             if not self._updater.is_retrieved_version_allowed_to_update(agent_family):
                 return
             self._updater.log_new_agent_update_message()
-            self._updater.purge_extra_agents_from_disk()
             agent = self._updater.download_and_get_new_agent(self._protocol, agent_family, goal_state)
-            if agent.is_blacklisted or not agent.is_downloaded:
-                msg = "Downloaded agent version is in bad state : {0} , skipping agent update".format(
-                    str(agent.version))
+
+            # We are not updating the agent if it is not downloaded properly or agent is blacklisted and already attempted 3 times.
+            # For some reason, if agent is blackslisted in previous attempts(<3), we clear the error state to make agent as good agent.
+            # If we allow to update, we increment the update attempt count.
+            if not agent.is_downloaded:
+                msg = "Agent: {0} is not downloaded properly, skipping agent update".format(str(agent.version))
```

Comment:
In the case of download failures, I don't want to continue with update and exit the current process. if download failure happens, we delete the directory. So, we could use directory presence for this check instead manifest since we don't use manifest anymore if that's your concern.

As far as telemetry, we do send telemetry for download failures, so checking both would give context.

Reply:
my concerns is that it seems like a useless check. if we are trying to check for download failures, shouldn't this check or similar be done when we download the agent? then the error message can have the appropriate context

Reply:
The download happens inside this GuestAgent module. ahent_update_handler only way know by calling is_download checks

`        agent = GuestAgent.from_agent_package(package_to_download, protocol, is_fast_track_goal_state)
`

Reply:
I'm confused here... we call self._updater.download_and_get_new_agent() on line 185 and then immediately check agent.is_downloaded.

Does that mean that download_and_get_new_agent() can fail to download the agent and not report the error with an exception? that does not seem correct to me.

Reply:
The GuestAgent class implemented in that way before and required refactoring of changing these methods and removing some of the dead code. That is separate effort I feel.